### ISSUE TITLE
Kubernetes support extension?

### ISSUE THREAD

**Issue by t3hmrman** [Mailu/Mailu#232]
**tldr;**
**- I am runing mailu on kubernetes, I started with the config but inter-container communication was broken, I made a new config but split it up into more deployments**
**- The new ingress resources were unused, it was unclear how to expose the mail server**
**- receiving email is still not working for me, but I'm working through it/debugging, would like some help, `anvil` seems to realize there's an email but not do anything about it).**

Hey all, thanks for all the great work on mailu, I just started using it (no more managing postfix/dovecot for me!), and am really happy with it. 

I am running Mailu on a kubernetes cluster and after combing through the resource configuration and replacing all necessary variables, I found out that there were some problems with communication between the containers. In particular the `smtpd_sasl_path` is normally set to `inet:imap:2102`, and this SHOULD find the local `imap` service, but I found that it would fail (timeout when you curl from inside the mailu pod), despite the fact that `imap` would be reachable from a testing container I created.

I got the mailu pod to work by replacing `inet:imap:2102` with `inet:<cluster ip of mailu pod>:2102`. I'm not 100% sure why the service didn't work (it should have, and did from other test containers), it might be related to [this issue about docker and alpine linux](https://forums.docker.com/t/resolved-service-name-resolution-broken-on-alpine-and-docker-1-11-1-cs1/19307/1). To fix it, I broke apart the pod into a few different deployments (this might be a bit against traditional kubernetes style, since you could think of mailu as one "app"), and added individual services and receiving mail seems to be working with no config changes (the `imap` service is working properly now).

Also, there is a lack of mention of the [Ingress Resources](https://kubernetes.io/docs/concepts/services-networking/ingress/) which I use to expose the mail server -- it requires the use of the generalized TCP loadbalancing (I've also included this, for those running their own nginx-based ingress controllers).

I'd like to post my config here for discussion/correction/etc, please find it below:

`mailu.yaml`
```yaml
---

apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: mailu-admin-ing
  labels:
    app: mailu
    role: mail
    tier: backend
spec:
  tls:
  - hosts:
    - "mail.example.com"
    secretName: letsencrypt-certs-all
  rules:
  - host: "mail.example.com"
    http:
      paths:
      - path: "/admin"
        backend:
          serviceName: mailu-admin
          servicePort: 80

---

apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: mailu-redis
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: mailu-redis
        role: mail
        tier: backend
    spec:
      containers:
      - name: redis
        image: redis:latest
        imagePullPolicy: Always
        volumeMounts:
          - mountPath: /data
            name: redisdata
        ports:
          - containerPort: 6379
            name: redis
            protocol: TCP
      volumes:
        - name: redisdata
          hostPath:
            path: /var/data/mailu/redisdata

---

apiVersion: v1
kind: Service
metadata:
  name: redis
  labels:
    app: mailu-redis
    role: mail
    tier: backend
spec:
  selector:
    app: mailu
    role: mail
    tier: backend
  ports:
  - name: redis
    port: 6379
    protocol: TCP

---

apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: mailu-imap
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: mailu-imap
        role: mail
        tier: backend
    spec:
      containers:
      - name: imap
        image: mailu/dovecot:stable
        imagePullPolicy: Always
        env:
          - name  : DOMAIN
            value : example.com
          - name  : HOSTNAME
            value : example.com
          - name  : POSTMASTER
            value : admin
        volumeMounts:
          - mountPath: /data
            name: maildata
          - mountPath: /mail
            name: mailstate
          - mountPath: /overrides
            name: overrides
          - mountPath: /certs
            name: certs
            readOnly: true
        ports:
          - containerPort: 2102
          - containerPort: 2525
          - containerPort: 143
          - containerPort: 993
          - containerPort: 4190
      volumes:
        - name: maildata
          hostPath:
            path: /var/data/mailu/maildata
        - name: mailstate
          hostPath:
            path: /var/data/mailu/mailstate
        - name: overrides
          hostPath:
            path: /var/data/mailu/overrides
        - name: certs
          secret:
            items:
              - key: tls.crt
                path: cert.pem
              - key: tls.key
                path: key.pem
            secretName: letsencrypt-certs-all

---

apiVersion: v1
kind: Service
metadata:
  name: imap
  labels:
    app: mailu
    role: mail
    tier: backend
spec:
  selector:
    app: mailu-imap
    role: mail
    tier: backend
  ports:
  ports:
  - name: imap-auth
    port: 2102
    protocol: TCP
  - name: imap-transport
    port: 2525
    protocol: TCP
  - name: imap-default
    port: 143
    protocol: TCP
  - name: imap-ssl
    port: 993
    protocol: TCP
  - name: sieve
    port: 4190
    protocol: TCP

---

apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: mailu-smtp
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: mailu-smtp
        role: mail
        tier: backend
    spec:
      containers:
      - name: smtp
        image: mailu/postfix:stable
        imagePullPolicy: Always
        env:
          - name  : DOMAIN
            value : example.com
          - name  : HOSTNAME
            value : mail.example.com
          - name  : MESSAGE_SIZE_LIMIT
            value : "50000000"
          - name  : RELAYHOST
            value : ""
        volumeMounts:
          - mountPath: /data
            name: maildata
          - mountPath: /overrides
            name: overrides
          - mountPath: /certs
            name: certs
            readOnly: true
        ports:
          - name: smtp
            containerPort: 25
            protocol: TCP
          - name: smtp-ssl
            containerPort: 465
            protocol: TCP
          - name: smtp-starttls
            containerPort: 587
            protocol: TCP
      volumes:
        - name: maildata
          hostPath:
            path: /var/data/mailu/maildata
        - name: overrides
          hostPath:
            path: /var/data/mailu/overrides
        - name: certs
          secret:
            items:
              - key: tls.crt
                path: cert.pem
              - key: tls.key
                path: key.pem
            secretName: letsencrypt-certs-all

---

apiVersion: v1
kind: Service
metadata:
  name: smtp
  labels:
    app: mailu
    role: mail
    tier: backend
spec:
  selector:
    app: mailu-smtp
    role: mail
    tier: backend
  ports:
  - name: smtp
    port: 25
    protocol: TCP
  - name: smtp-ssl
    port: 465
    protocol: TCP
  - name: smtp-starttls
    port: 587
    protocol: TCP

---

apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: mailu-milter
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: mailu-milter
        role: mail
        tier: backend
    spec:
      containers:
      - name: milter
        image: mailu/rmilter:stable
        imagePullPolicy: Always
        ports:
          - name: milter
            containerPort: 9900
            protocol: TCP
        volumeMounts:
          - name: maildata
            mountPath: /data
          - name: dkim
            mountPath: /dkim
          - name: overrides
            mountPath: /overrides
          - name: certs
            mountPath: /certs
            readOnly: true
      volumes:
        - name: maildata
          hostPath:
            path: /var/data/mailu/maildata
        - name: overrides
          hostPath:
            path: /var/data/mailu/overrides
        - name: dkim
          hostPath:
            path: /var/data/mailu/dkim
        - name: certs
          secret:
            items:
              - key: tls.crt
                path: cert.pem
              - key: tls.key
                path: key.pem
            secretName: letsencrypt-certs-all

---

apiVersion: v1
kind: Service
metadata:
  name: milter
  labels:
    app: mailu-milter
    role: mail
    tier: backend
spec:
  selector:
    app: mailu-milter
    role: mail
    tier: backend
  ports:
  - name: milter
    port: 9900
    protocol: TCP

---

apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: mailu-security
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: mailu-security
        role: mail
        tier: backend
    spec:
      containers:
      - name: antispam
        image: mailu/rspamd:stable
        imagePullPolicy: Always
        ports:
          - name: antispam
            containerPort: 11333
            protocol: TCP
        volumeMounts:
          - name: filter
            mountPath: /var/lib/rspamd
      - name: antivirus
        image: mailu/clamav:stable
        imagePullPolicy: Always
        ports:
          - name: antivirus
            containerPort: 3310
            protocol: TCP
        volumeMounts:
          - name: filter
            mountPath: /data
      volumes:
        - name: filter
          hostPath:
            path: /var/data/mailu/filter

---

apiVersion: v1
kind: Service
metadata:
  name: antispam
  labels:
    app: mailu-antispam
    role: mail
    tier: backend
spec:
  selector:
    app: mailu-security
    role: mail
    tier: backend
  ports:
  - name: antispam
    port: 11333
    protocol: TCP

---

apiVersion: v1
kind: Service
metadata:
  name: antivirus
  labels:
    app: mailu-antivirus
    role: mail
    tier: backend
spec:
  selector:
    app: mailu-security
    role: mail
    tier: backend
  ports:
  - name: antivirus
    port: 3310
    protocol: TCP

---

apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: mailu-admin
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: mailu-admin
        role: mail
        tier: backend
    spec:
      containers:
      - name: admin
        image: mailu/admin:stable
        imagePullPolicy: Always
        env:
          - name  : DOMAIN
            value : example.com
          - name  : HOSTNAME
            value : mail.example.com
          - name  : POSTMASTER
            value : core
          - name  : SECRET_KEY
            value : replace-this-with-a-secret
          - name  : DEBUG
            value : "True"
        volumeMounts:
          - name: maildata
            mountPath: /data
          - name: dkim
            mountPath: /dkim
          - name: certs
            mountPath: /certs
            readOnly: true
          # - name: docker
          #   mountPath: /var/run/docker.sock
          #   readOnly: true
        ports:
          - name: http
            containerPort: 80
            protocol: TCP
      volumes:
        - name: maildata
          hostPath:
            path: /var/data/mailu/maildata
        - name: dkim
          hostPath:
            path: /var/data/mailu/dkim
        - name: certs
          secret:
            items:
              - key: tls.crt
                path: cert.pem
              - key: tls.key
                path: key.pem
            secretName: letsencrypt-certs-all
        # - name: docker
        #   hostPath:
        #     path: /var/run/docker.sock

---

apiVersion: v1
kind: Service
metadata:
  name: mailu-admin
  labels:
    app: mailu-admin
    role: mail
    tier: backend
spec:
  selector:
    app: mailu-admin
    role: mail
    tier: backend
  ports:
  - name: http
    port: 80
    protocol: TCP
```

This config makes a heck of a lot more deployments, but my deployment of kubernetes at least manages to get the`imap` service right so I'm using it. Also if you want to delete all of it, you can still do a `kubectl delete -f <...>`.

I'm having problems receiving emails that I'm still trying to work through.

`nginx-ingress-controller.yaml`
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: nginx-ingress-conf
  namespace: kube-system
  labels:
    k8s-app: nginx-ingress-controller
data:
  25: "default/smtp:25"
  143: "default/imap:143"
  465: "default/smtp:465"
  587: "default/smtp:587"
  993: "default/imap:993"

---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx-ingress-controller
  labels:
    k8s-app: nginx-ingress-controller
  namespace: kube-system
spec:
  replicas: 1
  template:
    metadata:
      labels:
        k8s-app: nginx-ingress-controller
      annotations:
        prometheus.io/port: '10254'
        prometheus.io/scrape: 'true'
    spec:
      # hostNetwork makes it possible to use ipv6 and to preserve the source IP correctly regardless of docker configuration
      # however, it is not a hard dependency of the nginx-ingress-controller itself and it may cause issues if port 10254 already is taken on the host
      # that said, since hostPort is broken on CNI (https://github.com/kubernetes/kubernetes/issues/31307) we have to use hostNetwork where CNI is used
      # like with kubeadm
      # hostNetwork: true
      terminationGracePeriodSeconds: 60
      containers:
      - image: gcr.io/google_containers/nginx-ingress-controller:0.9.0-beta.11
        name: nginx-ingress-controller
        readinessProbe:
          httpGet:
            path: /healthz
            port: 10254
            scheme: HTTP
        livenessProbe:
          httpGet:
            path: /healthz
            port: 10254
            scheme: HTTP
          initialDelaySeconds: 10
          timeoutSeconds: 1
        ports:
        - containerPort: 80
          hostPort: 80
        - containerPort: 25
          hostPort: 25
        - containerPort: 143
          hostPort: 143
        - containerPort: 443
          hostPort: 443
        - containerPort: 465
          hostPort: 465
        - containerPort: 587
          hostPort: 587
        - containerPort: 993
          hostPort: 993
        env:
          - name: POD_NAME
            valueFrom:
              fieldRef:
                fieldPath: metadata.name
          - name: POD_NAMESPACE
            valueFrom:
              fieldRef:
                fieldPath: metadata.namespace
        args:
        - /nginx-ingress-controller
        - --default-backend-service=$(POD_NAMESPACE)/default-http-backend
        - --tcp-services-configmap=$(POD_NAMESPACE)/nginx-ingress-conf
```

Note the use of the `tcp-services-configmap` option. More information is available in [the official repo](https://github.com/kubernetes/ingress/tree/master/examples/tcp/nginx)

**Where I am now**
Right now, receiving emails to the mailu instance doesn't work for me (sending works). Not exactly sure why but I'm working on it, and will update the config here once I get it working.

**Comment by t3hmrman** [comment]
Silly me, I forgot about port 25 -- I've updated the config above to expose that port as well.

**Comment by kaiyou** [comment]
Okay, sorry for the delay and glad you figured it out.
If you are willing to contribute your script, feel free to fork and pull request after prefixing your files with `kubernetes-` :)
